flow360/component/simulation/unit_system.py

```diff
@@ -157,25 +157,39 @@ def _has_dimensions(quant, dim):
     return arg_dim == dim
 
 
-def _unit_object_parser(value, unyt_types: List[type]):
+def _unit_object_parser(value, unyt_types: List[type], **kwargs):
     """
     Parses {'value': value, 'units': units}, into unyt_type object : unyt.unyt_quantity, unyt.unyt_array
     """
-    if isinstance(value, dict) and "units" in value:
-        if "value" in value:
-            for unyt_type in unyt_types:
-                try:
-                    return unyt_type(value["value"], value["units"], dtype=np.float64)
-                except u.exceptions.UnitParseError:
-                    pass
-                except RuntimeError:
-                    pass
-                except KeyError:
-                    pass
-        else:
-            raise TypeError(
-                f"Dimensioned type instance {value} expects a 'value' field which was not given"
-            )
+    if isinstance(value, dict) is False or "units" not in value:
+        return value
+    if "value" not in value:
+        raise TypeError(
+            f"Dimensioned type instance {value} expects a 'value' field which was not given"
+        )
+    for unyt_type in unyt_types:
+        try:
+            dimensioned_value = unyt_type(value["value"], value["units"], dtype=np.float64)
+            # Check for Nan. This usually is a result of None being in the input value.
+            # with dtype=np.float64, None will be converted to np.nan.
+            # We need (IIRC) np.float64 to ensure consistent hashing
+            if kwargs.get("allow_inf_nan", False) is False:
+                if np.ndim(dimensioned_value.value) == 0 and np.isnan(dimensioned_value.value):
+                    raise ValueError("NaN or None found in input scalar which is not allowed.")
+                if np.ndim(dimensioned_value.value) > 0 and any(np.isnan(dimensioned_value.value)):
+                    raise ValueError("NaN or None found in input array which is not allowed.")
+            return dimensioned_value
```

Comment:
can we put this logic to a separate validator and keep this validator only a parser? See the logic of `_VectorType.validate()` it checks step by step all conditions, it is easier to follow.

Reply:
Done